pm/recipient_test.go

```diff
@@ -0,0 +1,436 @@
+	}
+
+	_, err = r.ReceiveTicket(ticket, sig, params.Seed)
+	if err == nil {
+		t.Error("expected invalid recipientRand revealed error")
+	}
+	if err != nil && err != errInvalidRecipientRandRevealed {
+		t.Errorf("expected invalid recipientRand revealed error, got %v", err)
+	}
+
+	// Test invalid senderNonce
+	params, err = r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Receive senderNonce = 0
+	ticket0 := &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            sender,
+		FaceValue:         faceValue,
+		WinProb:           winProb,
+		SenderNonce:       0,
+		RecipientRandHash: params.RecipientRandHash,
+	}
+
+	_, err = r.ReceiveTicket(ticket0, sig, params.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Receive senderNonce = 1
+	ticket1 := &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            sender,
+		FaceValue:         faceValue,
+		WinProb:           winProb,
+		SenderNonce:       1,
+		RecipientRandHash: params.RecipientRandHash,
+	}
+
+	_, err = r.ReceiveTicket(ticket1, sig, params.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Replay senderNonce = 1 (new nonce = highest seen nonce)
+	_, err = r.ReceiveTicket(ticket1, sig, params.Seed)
+	if err == nil {
+		t.Error("expected invalid senderNonce (new nonce = highest seen nonce) error")
+	}
+	if err != nil && err != errInvalidTicketSenderNonce {
+		t.Errorf("expected invalid senderNonce (new nonce = highest seen nonce) error, got %v", err)
+	}
+
+	// Replay senderNonce = 0 (new nonce < highest seen nonce)
+	_, err = r.ReceiveTicket(ticket0, sig, params.Seed)
+	if err == nil {
+		t.Error("expected invalid senderNonce (new nonce < highest seen nonce) error")
+	}
+	if err != nil && err != errInvalidTicketSenderNonce {
+		t.Errorf("expected invalid senderNonce (new nonce < highest seen nonce) error, got %v", err)
+	}
+}
+
+func TestRedeemWinningTicket(t *testing.T) {
+	sender := ethcommon.HexToAddress("A69cdA26600c155cF2c150964Bdb5371ac3f606F")
+	b := newStubBroker()
+	v := &stubValidator{}
+	ts := newStubTicketStore()
+	secret := [32]byte{3}
+	faceValue := big.NewInt(100)
+	winProb := big.NewInt(100)
+	sig := []byte("foo")
+
+	r := NewRecipient(b, v, ts, secret, faceValue, winProb)
+
+	params1, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub broker with zero deposit and penalty escrow
+	b.SetDeposit(sender, big.NewInt(0))
+	b.SetPenaltyEscrow(sender, big.NewInt(0))
+
+	// Test zero deposit and penalty escrow error
+	ticket := &Ticket{
+		Recipient:         ethcommon.Address{},
+		Sender:            sender,
+		FaceValue:         faceValue,
+		WinProb:           winProb,
+		SenderNonce:       0,
+		RecipientRandHash: params1.RecipientRandHash,
+	}
+
+	won, err := r.ReceiveTicket(ticket, sig, params1.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !won {
+		t.Fail()
```

Comment:
I'm not sure why this should be a different kind of failure compared to `err != nil`. I think both cases should be the same, probably using `t.Fatal` to stop execution, since in this specific test we're assuming that `ReceiveTicket` should work as expected, i.e. `won` should be true and `err` should be nil and anything different means we can't reliably test `RedeemWinningTicket`.

wdyt?

Reply:
Yeah in addition to maintaining consistent with method usage, I think it makes sense to add a log message with `Fatal()` for this case so there is at least a clear reason why the test is terminating execution instead of proceeding.